### ISSUE TITLE
Will tensorflow return the best model as a result of training

### ISSUE THREAD

**Issue by slevental** [aws/sagemaker-python-sdk#48]
I'm trying to understand the strategy of model evaluation implemented in TF within a container, my goal is to sage the most accurate model, not the most recent one. Is it possible via this API somehow?

**Comment by owen-t** [comment]
The TensorFlow container saves the most recent exported model. We provide TensorFlow checkpoints, which contain older models checkpoints produced during training.

**Comment by slevental** [comment]
@owen-t so, right now I should manually evaluate all checkpoints, find the best one and create SageMaker model + endpoint based on that. That's something that makes sense to do in the SageMaker library, what do you think?

**Comment by owen-t** [comment]
Agreed - allowing better control over model export in TensorFlow would be useful.

**Comment by chinazm** [comment]
Using training hooks and evaluation hooks can do both best model saving and early stopping.

EstimatorSpec(
            mode=spec.mode,
            predictions=spec.predictions,
            loss=spec.loss,
            train_op=spec.train_op,
            eval_metric_ops=spec.eval_metric_ops,
            export_outputs=spec.export_outputs,
            training_chief_hooks=spec.training_chief_hooks,
            training_hooks=training_hooks,
            scaffold=spec.scaffold,
            evaluation_hooks=evaluation_hooks)

**Comment by khu834** [comment]
> Using training hooks and evaluation hooks can do both best model saving and early stopping.
> 
> EstimatorSpec(
> mode=spec.mode,
> predictions=spec.predictions,
> loss=spec.loss,
> train_op=spec.train_op,
> eval_metric_ops=spec.eval_metric_ops,
> export_outputs=spec.export_outputs,
> training_chief_hooks=spec.training_chief_hooks,
> training_hooks=training_hooks,
> scaffold=spec.scaffold,
> evaluation_hooks=evaluation_hooks)

Do you have more specific examples or links that better demonstrate how to utilize training and evaluation hooks? e,g., how would one construct a tf.train.SessionRunHook that can read the evaluation metrics from inside model_fn over many iterations and compare the loss

I assume the EstimatorSpec is returned by model_fn?
Where is the 'spec' object here coming from?
Thank you.

**Comment by chinazm** [comment]
I would refer to this feature request:
https://github.com/aws/sagemaker-tensorflow-container/issues/75

**Comment by laurenyu** [comment]
Closing due to inactivity and the introduction of script mode with our TensorFlow containers. Script mode allows for greater flexibility in writing the TF training script, which should allow for using the hooks described above. For more information about script mode, see [our TF README](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/tensorflow/README.rst).
